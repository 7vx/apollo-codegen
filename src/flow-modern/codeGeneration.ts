import { parse } from 'babylon';
import generate from 'babel-generator';
import * as t from 'babel-types';
import { stripIndent } from 'common-tags';
import {
  getNamedType,
  GraphQLEnumType,
  GraphQLField,
  GraphQLInputObjectType,
  GraphQLInputField,
  GraphQLList,
  GraphQLObjectType,
  GraphQLNonNull,
  isCompositeType,
  GraphQLError,
  GraphQLType
} from 'graphql';

import {
  CompilerContext,
  Operation,
  Fragment,
  SelectionSet,
  Field,
} from '../compiler';

import {
  typeCaseForSelectionSet,
  Variant
} from '../compiler/visitors/typeCase';

// import {
//   collectFragmentsReferenced
// } from '../compiler/visitors/collectFragmentsReferenced';

import {
  collectAndMergeFields
} from '../compiler/visitors/collectAndMergeFields';

import { typeAnnotationFromGraphQLType } from './helpers';
import FlowGenerator, {
  ObjectProperty,
} from './language';
import Printer from './printer';

export function generateSource(
  context: CompilerContext,
  outputIndividualFiles?: boolean = false,
  only?: string,
) {
  const generator = new FlowAPIGenerator(context);

  if (outputIndividualFiles) {
    // TODO
  } else{
    generator.fileHeader();

    context.typesUsed
      .filter(type => type instanceof GraphQLEnumType)
      .forEach((enumType: GraphQLEnumType) => {
        generator.typeAliasForEnumType(enumType);
      });

    context.typesUsed
      .filter(type => type instanceof GraphQLInputObjectType)
      .forEach((enumType: GraphQLInputObjectType) => {
        generator.typeAliasForInputObjectType(enumType);
      });

    Object.values(context.operations).forEach(operation => {
      // generator.typeVariablesDeclarationForOperation(operation);
      generator.typeAliasesForOperation(operation);
    });

    Object.values(context.fragments).forEach(fragment => {
      // console.log('Fragment', fragment);
      generator.typeAliasesForFragment(fragment);
    });
  }

  return generator.output;
}

export class FlowAPIGenerator extends FlowGenerator {
  context: CompilerContext
  printer: Printer
  scopeStack: string[]

  constructor(context: CompilerContext) {
    super();

    this.context = context;
    this.printer = new Printer();
    this.scopeStack = [];
  }

  fileHeader() {
    this.printer.enqueue(
      stripIndent`
        /* @flow */
        // This file was automatically generated and should not be edited.
      `
    );
  }

  public typeAliasForEnumType(enumType: GraphQLEnumType) {
    this.printer.enqueue(this.enumerationDeclaration(enumType));
  }

  public typeAliasForInputObjectType(inputObjectType: GraphQLInputObjectType) {
    this.printer.enqueue(this.inputObjectDeclaration(inputObjectType));
  }

  public typeAliasesForOperation(operation: Operation) {
    const {
      operationType,
      operationName,
      variables,
      selectionSet
    } = operation;

    this.scopeStackPush(operationName);

    this.printer.enqueue(stripIndent`
      // ====================================================
      // GraphQL ${operationType} operation: ${operationName}
      // ====================================================
    `)

    // The root operation only has one variant
    // Do we need to get exhaustive variants anyway?
    const variants = this.getVariantsForSelectionSet(selectionSet);

    const variant = variants[0];
    const properties = this.getPropertiesForVariant(variant);

    const exportedTypeAlias = this.exportDeclaration(
      this.typeAliasObject(operationName, properties)
    );

    this.printer.enqueue(exportedTypeAlias);
    this.scopeStackPop();
  }

  public typeAliasesForFragment(fragment: Fragment) {
    const {
      fragmentName,
      type,
      selectionSet
    } = fragment;

    this.scopeStackPush(fragmentName);

    this.printer.enqueue(stripIndent`
      // ====================================================
      // GraphQL fragment: ${fragmentName}
      // ====================================================
    `);

    const variants = this.getVariantsForSelectionSet(selectionSet);

    const variant = variants[0];
    const properties = this.getPropertiesForVariant(variant);

    const exportedTypeAlias = this.exportDeclaration(
      this.typeAliasObject(fragmentName, properties)
    );

    this.printer.enqueue(exportedTypeAlias);
    this.scopeStackPop();
  }

  private getVariantsForSelectionSet(selectionSet: SelectionSet) {
    return this.getTypeCasesForSelectionSet(selectionSet).exhaustiveVariants;
  }

  private getTypeCasesForSelectionSet(selectionSet: SelectionSet) {
    return typeCaseForSelectionSet(
      selectionSet,
      this.context.options.mergeInFieldsFromFragmentSpreads
    );
  }

  private getPropertiesForVariant(variant: Variant): ObjectProperty[] {
    const fields = collectAndMergeFields(
      variant,
      this.context.options.mergeInFieldsFromFragmentSpreads
    );

    return fields.map(field => {
      this.scopeStackPush(field.name);

      let res;
      if (field.selectionSet) {
        const genericAnnotation = this.annotationFromScopeStack(this.scopeStack);
        res = this.handleFieldSelectionSetValue(
          genericAnnotation,
          field
        );
      } else {
        res = this.handleFieldValue(
          field,
          variant
        );
      }

      this.scopeStackPop();
      return res;
    });
  }

  private handleFieldSelectionSetValue(genericAnnotation: t.GenericTypeAnnotation, field: Field) {
    const { selectionSet } = field;

    const typeCase = this.getTypeCasesForSelectionSet(selectionSet as SelectionSet);
    const variants = typeCase.exhaustiveVariants;

    let exportedTypeAlias;
    if (variants.length === 1) {
      const variant = variants[0];
      const properties = this.getPropertiesForVariant(variant);
      exportedTypeAlias = this.exportDeclaration(
        this.typeAliasObject(genericAnnotation.id.name, properties)
      );
    } else {
      const propertySets = variants.map(variant => {
        return this.getPropertiesForVariant(variant);
      })

      exportedTypeAlias = this.exportDeclaration(
        this.typeAliasObjectUnion(
          genericAnnotation.id.name,
          propertySets
        )
      );
    }

    this.printer.enqueue(exportedTypeAlias);

    return {
      name: field.name,
      annotation: genericAnnotation
    };
  }

  private handleFieldValue(field: Field, variant: Variant) {
    let res;
    if (field.name === '__typename') {
      const annotations = variant.possibleTypes
        .map(type => {
          const annotation = t.stringLiteralTypeAnnotation();
          annotation.value = type.toString();
          return annotation;
        });

      res = {
        name: field.name,
        annotation: t.unionTypeAnnotation(annotations)
      };
    } else {
      // TODO: Double check that this works
      res = {
        name: field.name,
        annotation: typeAnnotationFromGraphQLType(field.type)
      };
    }

    return res;
  }

  public get output() {
    return this.printer.print();
  }

  scopeStackPush(name: string) {
    this.scopeStack.push(name);
  }

  scopeStackPop() {
    const popped = this.scopeStack.pop()
  }

}
