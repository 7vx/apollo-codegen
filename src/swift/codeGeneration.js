import {
  GraphQLError,
  getNamedType,
  isCompositeType,
  isAbstractType,
  isEqualType,
  GraphQLScalarType,
  GraphQLEnumType,
  GraphQLList,
  GraphQLNonNull,
  GraphQLID,
  GraphQLInputObjectType
} from 'graphql'

import  { isTypeProperSuperTypeOf } from '../utilities/graphql'

import {
  join,
  wrap,
} from '../utilities/printing';

import {
  namespaceDeclaration,
  classDeclaration,
  structDeclaration,
  propertyDeclaration,
  propertyDeclarations,
  escapeIdentifierIfNeeded
} from './language';

import {
  structNameForPropertyName,
  structNameForFragmentName,
  structNameForInlineFragment,
  operationClassName,
  enumCaseName,
  propertiesFromSelectionSet,
  propertyFromField,
} from './naming';

import {
  escapedString,
  multilineString,
  dictionaryLiteralForFieldArguments,
} from './values';

import {
  possibleTypesForType,
  typeNameFromGraphQLType,
  fieldTypeEnum,
} from './types';

import CodeGenerator from '../utilities/CodeGenerator';

export function generateSource(context) {
  const generator = new CodeGenerator(context);

  generator.printOnNewline('//  This file was automatically generated and should not be edited.');
  generator.printNewline();
  generator.printOnNewline('import Apollo');

  namespaceDeclaration(generator, context.namespace, () => {

    context.typesUsed.forEach(type => {
      typeDeclarationForGraphQLType(generator, type);
    });

    Object.values(context.operations).forEach(operation => {
      classDeclarationForOperation(generator, { ...operation, selectionSet: selectionSetFrom(operation) });
    });

    Object.values(context.fragments).forEach(fragment => {
      structDeclarationForFragment(generator, { ...fragment, selectionSet: selectionSetFrom(fragment) });
    });
  })

  return generator.output;
}

function selectionSetFrom({ fields, inlineFragments, fragmentSpreads }) {
  let selectionSet = [];

  fields && fields.forEach(field => {
    selectionSet.push({ kind: 'Field', ...field, selectionSet: selectionSetFrom(field) });
  });

  inlineFragments && inlineFragments.forEach(inlineFragment => {
    selectionSet.push({ kind: 'InlineFragment', ...inlineFragment, selectionSet: selectionSetFrom(inlineFragment) });
  });

  fragmentSpreads && fragmentSpreads.forEach(fragmentSpread => {
    selectionSet.push({ kind: 'FragmentSpread', fragmentName: fragmentSpread });
  });

  return selectionSet;
}

export function classDeclarationForOperation(
  generator,
  {
    operationName,
    operationType,
    rootType,
    variables,
    selectionSet,
    fragmentsReferenced,
    source,
  }
) {
  let className;
  let protocol;

  switch (operationType) {
    case 'query':
      className = `${operationClassName(operationName)}Query`;
      protocol = 'GraphQLQuery';
      break;
    case 'mutation':
      className = `${operationClassName(operationName)}Mutation`;
      protocol = 'GraphQLMutation';
      break;
    default:
      throw new GraphQLError(`Unsupported operation type "${operationType}"`);
  }

  classDeclaration(generator, {
    className,
    modifiers: ['public', 'final'],
    adoptedProtocols: [protocol]
  }, () => {
    if (source) {
      generator.printOnNewline('public static let operationString =');
      generator.withIndent(() => {
        multilineString(generator, source);
      });
    }

    if (fragmentsReferenced && fragmentsReferenced.length > 0) {
      generator.printOnNewline('public static var requestString: String { return operationString');
      fragmentsReferenced.forEach(fragment => {
        generator.print(`.appending(${structNameForFragmentName(fragment)}.fragmentString)`)
      });
      generator.print(' }');
    }

    generator.printNewlineIfNeeded();

    if (variables && variables.length > 0) {
      const properties = variables.map(({ name, type }) => {
        const propertyName = escapeIdentifierIfNeeded(name);
        const typeName = typeNameFromGraphQLType(generator.context, type);
        const isOptional = !(type instanceof GraphQLNonNull || type.ofType instanceof GraphQLNonNull);
        return { name, propertyName, type, typeName, isOptional };
      });

      propertyDeclarations(generator, properties);

      generator.printNewlineIfNeeded();
      initializerDeclarationForProperties(generator, properties);

      generator.printNewlineIfNeeded();
      generator.printOnNewline(`public var variables: GraphQLMap?`);
      generator.withinBlock(() => {
        generator.printOnNewline(wrap(
          `return [`,
          join(properties.map(({ name, propertyName }) => `"${name}": ${propertyName}`), ', ') || ':',
          `]`,
        ));
      });
    } else {
      initializerDeclarationForProperties(generator, []);
    }

    structDeclarationForSelectionSet(
      generator,
      {
        structName: "Data",
        parentType: rootType,
        selectionSet
      }
    );
  });
}

export function structDeclarationForFragment(
  generator,
  {
    fragmentName,
    typeCondition,
    selectionSet,
    source
  }
) {
  const structName = structNameForFragmentName(fragmentName);

  structDeclarationForSelectionSet(generator, {
    structName,
    adoptedProtocols: ['GraphQLFragment'],
    parentType: typeCondition,
    selectionSet
  }, () => {
    if (source) {
      generator.printOnNewline('public static let fragmentString =');
      generator.withIndent(() => {
        multilineString(generator, source);
      });
    }
  });
}

export function structDeclarationForSelectionSet(
  generator,
  {
    structName,
    adoptedProtocols = ['GraphQLSelectionSet'],
    parentType,
    selectionSet
  },
  beforeClosure
) {
  const possibleTypes = parentType ? possibleTypesForType(generator.context, parentType) : null;

  structDeclaration(generator, { structName, adoptedProtocols }, () => {
    if (beforeClosure) {
      beforeClosure();
    }

    if (possibleTypes) {
      generator.printNewlineIfNeeded();
      generator.printOnNewline('public static let possibleTypes = [');
      generator.print(join(possibleTypes.map(type => `"${String(type)}"`), ', '));
      generator.print(']');
    }

    generator.printNewlineIfNeeded();
    generator.printOnNewline('public static let selections: [Selection] = ');
    selectionSetInitialization(generator, selectionSet, structName);

    generator.printNewlineIfNeeded();

    propertyDeclaration(generator, { propertyName: "snapshot", typeName: "Snapshot" });

    generator.printNewlineIfNeeded();
    generator.printOnNewline('public init(snapshot: Snapshot)');
    generator.withinBlock(() => {
      generator.printOnNewline(`self.snapshot = snapshot`);
    });

    const properties = propertiesFromSelectionSet(generator.context, selectionSet);

    const fields = properties.filter(property => property.kind === 'Field' && property.propertyName !== '__typename');
    const inlineFragments = properties.filter(property => property.kind === 'InlineFragment');
    const fragmentSpreads = properties.filter(property => property.kind === 'FragmentSpread');

    if (!possibleTypes || possibleTypes.length == 1) {
      generator.printNewlineIfNeeded();
      generator.printOnNewline(`public init`);
      generator.print('(');
      generator.print(join(fields.map(({ propertyName, type, typeName, isOptional }) =>
        join([
          `${propertyName}: ${typeName}`,
          isOptional && ' = nil'
        ])
      ), ', '));
      generator.print(')');

      generator.withinBlock(() => {
        generator.printOnNewline(wrap(
          `self.init(snapshot: [`,
          join([
            possibleTypes && possibleTypes.length > 0 && `"__typename": "${possibleTypes[0]}"`,
            ...fields.map(({ name, propertyName }) => `"${propertyName}": ${propertyName}`)
          ], ', ') || ':',
          `])`
        ));
      });
    } else {
      possibleTypes.forEach(possibleType => {
        generator.printNewlineIfNeeded();
        generator.printOnNewline(`public static func make${possibleType}`);
        generator.print('(');

        const inlineFragment = inlineFragments.find(inlineFragment => inlineFragment.typeCondition === possibleType);
        let properties;
        if (inlineFragment) {
          properties = propertiesFromSelectionSet(generator.context, inlineFragment.selectionSet, inlineFragment.structName);
        } else {
          properties = propertiesFromSelectionSet(generator.context, selectionSet);
        }
        const fields = properties.filter(property => property.kind === 'Field' && property.propertyName !== '__typename');

        generator.print(join(fields.map(({ propertyName, type, typeName, isOptional }) =>
          join([
            `${propertyName}: ${typeName}`,
            isOptional && ' = nil'
          ])
        ), ', '));
        generator.print(`) -> ${structName}`);

        generator.withinBlock(() => {
          generator.printOnNewline(wrap(
            `return ${structName}(snapshot: [`,
            join([
              `"__typename": "${possibleType}"`,
              ...fields.map(({ name, propertyName }) => `"${propertyName}": ${propertyName}`)
            ], ', '),
            `])`
          ));
        });
      });
    }

    properties.forEach(property => {
      if (property.kind === 'FragmentSpread') return;

      generator.printNewlineIfNeeded();
      propertyDeclarationForSelection(generator, property);
    });

    if (fragmentSpreads.length > 0) {
      generator.printNewlineIfNeeded();
      generator.printOnNewline(`public var fragments: Fragments`);
      generator.withinBlock(() => {
        generator.printOnNewline("get");
        generator.withinBlock(() => {
          generator.printOnNewline(`return Fragments(snapshot: snapshot)`);
        });
        generator.printOnNewline("set");
        generator.withinBlock(() => {
          generator.printOnNewline(`snapshot = newValue.snapshot`);
        });
      });
    }

    if (inlineFragments.length > 0) {
      inlineFragments.forEach(({ structName, typeCondition, selectionSet }) => {
        structDeclarationForSelectionSet(
          generator,
          {
            structName: structName,
            parentType: typeCondition,
            possibleTypes: possibleTypesForType(generator.context, typeCondition),
            adoptedProtocols: ['GraphQLFragment'],
            selectionSet: selectionSet
          }
        );
      });
    }

    if (fragmentSpreads.length > 0) {
      structDeclaration(
        generator,
        {
          structName: 'Fragments'
        },
        () => {
          propertyDeclaration(generator, { propertyName: "snapshot", typeName: "Snapshot" });
          fragmentSpreads.forEach(({ propertyName, bareTypeName, typeName, fragment })  => {
            const isOptional = !isTypeProperSuperTypeOf(generator.context.schema, fragment.typeCondition, parentType);

            generator.printNewlineIfNeeded();
            generator.printOnNewline(`public var ${propertyName}: ${isOptional ? typeName + '?' : typeName}`);
            generator.withinBlock(() => {
              generator.printOnNewline("get");
              generator.withinBlock(() => {
                if (isOptional) {
                  generator.printOnNewline(`if !${typeName}.possibleTypes.contains(snapshot["__typename"]! as! String) { return nil }`);
                }
                generator.printOnNewline(`return ${typeName}(snapshot: snapshot)`);
              });
              generator.printOnNewline("set");
              generator.withinBlock(() => {
                if (isOptional) {
                  generator.printOnNewline(`guard let newValue = newValue else { return }`);
                  generator.printOnNewline(`snapshot = newValue.snapshot`);
                } else {
                  generator.printOnNewline(`snapshot = newValue.snapshot`);
                }
              });
            });
          })
        }
      );
    }

    fields.filter(field => isCompositeType(getNamedType(field.type))).forEach(field => {
      structDeclarationForSelectionSet(
        generator,
        {
          structName: structNameForPropertyName(field.responseName),
          parentType: getNamedType(field.type),
          selectionSet: field.selectionSet
        }
      );
    });
  });
}

function mapExpressionForType(context, type, expression, prefix = '') {
  let isOptional;
  if (type instanceof GraphQLNonNull) {
    isOptional = false;
    type = type.ofType;
  } else {
    isOptional = true;
  }

  if (type instanceof GraphQLList) {
    if (isOptional) {
      return `${prefix}.flatMap { $0.map { ${mapExpressionForType(context, type.ofType, expression, '$0')} } }`;
    } else {
      return `${prefix}.map { ${mapExpressionForType(context, type.ofType, expression, '$0')} }`;
    }
  } else if (isOptional) {
    return `${prefix}.flatMap { ${expression} }`;
  } else {
    return expression;
  }
}

function propertyDeclarationForSelection(generator, selection) {
  const { kind, propertyName, typeName, type, isConditional, description } = selection;

  generator.printOnNewline(description && ` /// ${description}`);
  generator.printOnNewline(`public var ${propertyName}: ${typeName}`);
  generator.withinBlock(() => {
    const namedType = getNamedType(type);

    if (kind === 'InlineFragment') {
      const structName = structNameForInlineFragment(selection);

      generator.printOnNewline("get");
      generator.withinBlock(() => {
        generator.printOnNewline(`if !${structName}.possibleTypes.contains(__typename) { return nil }`);
        generator.printOnNewline(`return ${structName}(snapshot: snapshot)`);
      });
      generator.printOnNewline("set");
      generator.withinBlock(() => {
        generator.printOnNewline(`guard let newValue = newValue else { return }`);
        generator.printOnNewline(`snapshot = newValue.snapshot`);
      });
    } else if (isCompositeType(namedType)) {
      const isOptional = isConditional || !(type instanceof GraphQLNonNull);
      const isList = type instanceof GraphQLList || type.ofType instanceof GraphQLList;
      const structName = escapeIdentifierIfNeeded(structNameForPropertyName(propertyName));

      if (isList) {
        generator.printOnNewline("get");
        generator.withinBlock(() => {
          const snapshotTypeName = typeNameFromGraphQLType(generator.context, type, 'Snapshot', isOptional);
          let getter = `return (snapshot["${propertyName}"]! as! ${snapshotTypeName})`;
          getter += mapExpressionForType(generator.context, type, `${structName}(snapshot: $0)`);
          generator.printOnNewline(getter);
        });
        generator.printOnNewline("set");
        generator.withinBlock(() => {
          let newValueExpression = "newValue" + mapExpressionForType(generator.context, type, `$0.snapshot`);
          generator.printOnNewline(`snapshot.updateValue(${newValueExpression}, forKey: "${propertyName}")`);

        });
      } else {
        generator.printOnNewline("get");
        generator.withinBlock(() => {
          generator.printOnNewline(`return ${structName}(snapshot: snapshot["${propertyName}"]! as! Snapshot)`);
        });
        generator.printOnNewline("set");
        generator.withinBlock(() => {
          let newValueExpression;
          if (isOptional) {
            newValueExpression = 'newValue?.snapshot';
          } else {
            newValueExpression = 'newValue.snapshot';
          }
          generator.printOnNewline(`snapshot.updateValue(${newValueExpression}, forKey: "${propertyName}")`);
        });
      }
    } else {
      generator.printOnNewline("get");
      generator.withinBlock(() => {
        generator.printOnNewline(`return snapshot["${propertyName}"]! as! ${typeName}`);
      });
      generator.printOnNewline("set");
      generator.withinBlock(() => {
        generator.printOnNewline(`snapshot.updateValue(newValue, forKey: "${propertyName}")`);
      });
    }
  });
}

export function initializerDeclarationForProperties(generator, properties) {
  generator.printOnNewline(`public init`);
  generator.print('(');
  generator.print(join(properties.map(({ propertyName, type, typeName, isOptional }) =>
    join([
      `${propertyName}: ${typeName}`,
      isOptional && ' = nil'
    ])
  ), ', '));
  generator.print(')');

  generator.withinBlock(() => {
    properties.forEach(({ propertyName }) => {
      generator.printOnNewline(`self.${propertyName} = ${propertyName}`);
    });
  });
}

export function selectionSetInitialization(generator, selectionSet, parentStructName) {
  generator.print('[');
  generator.withIndent(() => {
    selectionSet.forEach(selection => {
      if (selection.kind === 'Field') {
        const { responseName, fieldName, args, type } = selection;
        const structName = join([parentStructName, structNameForPropertyName(responseName)], '.');

        generator.printOnNewline(`Field(`);
        generator.print(join([
          `"${fieldName}"`,
          responseName != fieldName ? `alias: "${responseName}"` : null,
          args && args.length && `arguments: ${dictionaryLiteralForFieldArguments(args)}`,
          `type: ${fieldTypeEnum(generator.context, type, structName)}`
        ], ', '));
        generator.print('),');
      } else if (selection.kind === 'FragmentSpread') {
        const structName = structNameForFragmentName(selection.fragmentName);
        generator.printOnNewline(`FragmentSpread(${structName}.self),`);
      } else if (selection.kind === 'InlineFragment') {
        const structName = join([parentStructName, structNameForInlineFragment(selection)], '.');
        generator.printOnNewline(`FragmentSpread(${structName}.self),`);
      }
    });
  });
  generator.printOnNewline(']');
}

export function typeDeclarationForGraphQLType(generator, type) {
  if (type instanceof GraphQLEnumType) {
    enumerationDeclaration(generator, type);
  } else if (type instanceof GraphQLInputObjectType) {
    structDeclarationForInputObjectType(generator, type);
  }
}

function enumerationDeclaration(generator, type) {
  const { name, description } = type;
  const values = type.getValues();

  generator.printNewlineIfNeeded();
  generator.printOnNewline(description && `/// ${description}`);
  generator.printOnNewline(`public enum ${name}: String`);
  generator.withinBlock(() => {
    values.forEach(value =>
      generator.printOnNewline(`case ${escapeIdentifierIfNeeded(enumCaseName(value.name))} = "${value.value}"${wrap(' /// ', value.description)}`)
    );
  });
  generator.printNewline();
  generator.printOnNewline(`extension ${name}: JSONDecodable, JSONEncodable {}`);
}

function structDeclarationForInputObjectType(generator, type) {
  const { name: structName, description } = type;
  const adoptedProtocols = ['GraphQLMapConvertible'];
  const fields = Object.values(type.getFields());
  const properties = fields.map(field => propertyFromField(generator.context, field));

  structDeclaration(generator, { structName, description, adoptedProtocols }, () => {
    generator.printOnNewline(`public var graphQLMap: GraphQLMap`);

    generator.printNewlineIfNeeded();
    generator.printOnNewline(`public init`);
    generator.print('(');
    generator.print(join(properties.map(({ propertyName, type, typeName, isOptional }) => {
      if (isOptional) {
        return `${propertyName}: Optional<${typeName}> = nil`;
      } else {
        return `${propertyName}: ${typeName}`;
      }
    }), ', '));
    generator.print(')');

    generator.withinBlock(() => {
      generator.printOnNewline(wrap(
        `graphQLMap = [`,
        join(properties.map(({ name, propertyName }) => `"${name}": ${propertyName}`), ', ') || ':',
        `]`
      ));
    });
  });
}
