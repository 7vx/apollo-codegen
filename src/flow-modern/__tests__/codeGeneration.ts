import { parse, GraphQLNonNull, GraphQLString, GraphQLEnumType, GraphQLList } from 'graphql';

import { loadSchema } from '../../loading';
const schema = loadSchema(require.resolve('../../../test/fixtures/starwars/schema.json'));

import {
  compileToIR,
  CompilerOptions,
  CompilerContext,
  SelectionSet,
  Field,
  Argument
} from '../../compiler';

import { generateSource } from '../codeGeneration';

function compile(
  source: string,
  options: CompilerOptions = {
    mergeInFieldsFromFragmentSpreads: true,
    addTypename: true
  }
): CompilerContext {
  const document = parse(source);
  return compileToIR(schema, document, options);
}

describe('Flow codeGeneration', () => {
  test('simple hero query', () => {
    const context = compile(`
      query HeroName($episode: Episode) {
        hero(episode: $episode) {
          name
          id
        }
      }
    `);
    const output = generateSource(context);
    expect(output).toMatchSnapshot();
  });

  test('inline fragments', () => {
    const context = compile(`
      query HeroName($episode: Episode) {
        hero(episode: $episode) {
          name
          id

          ... on Human {
            homePlanet
            friends {
              name
            }
          }

          ... on Droid {
            appearsIn
          }
        }
      }
    `);
    const output = generateSource(context);
    expect(output).toMatchSnapshot();
  });

  test('fragment spreads', () => {
    const context = compile(`
      query HeroName($episode: Episode) {
        hero(episode: $episode) {
          name
          id
          ...humanFragment
          ...droidFragment
        }
      }

      fragment humanFragment on Human {
        homePlanet
        friends {
          ... on Human {
            name
          }

          ... on Droid {
            id
          }
        }
      }

      fragment droidFragment on Droid {
        appearsIn
      }
    `);
    const output = generateSource(context);
    expect(output).toMatchSnapshot();
  });

  test.only(`should generate mutation operations with complex input types`, function() {
    const context = compile(`
      mutation ReviewMovie($episode: Episode, $review: ReviewInput) {
        createReview(episode: $episode, review: $review) {
          stars
          commentary
        }
      }
    `);

    const output = generateSource(context);
    console.log(output);
    expect(output).toMatchSnapshot();
  });
});
