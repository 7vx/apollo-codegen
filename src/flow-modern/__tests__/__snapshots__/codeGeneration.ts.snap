// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Flow codeGeneration fragment spreads with inline fragments 1`] = `
"

/* @flow */
// This file was automatically generated and should not be edited.

// The episodes in the Star Wars trilogy
export type Episode = \\"NEWHOPE\\" | \\"EMPIRE\\" | \\"JEDI\\";

// ====================================================
// GraphQL query operation: HeroName
// ====================================================

export type HeroName_hero_friends = {|
  __typename: \\"Human\\",
  // What this human calls themselves
  name: string,
|} | {|
  __typename: \\"Droid\\",
  // The ID of the droid
  id: string,
|};

export type HeroName_hero = {|
  __typename: \\"Human\\",
  // What this human calls themselves
  name: string,
  // The ID of the human
  id: string,
  // The home planet of the human, or null if unknown
  homePlanet: ?string,
  // This human's friends, or an empty list if they have none
  friends: HeroName_hero_friends,
|} | {|
  __typename: \\"Droid\\",
  // What others call this droid
  name: string,
  // The ID of the droid
  id: string,
  // The movies this droid appears in
  appearsIn: (?Episode)[],
|};

export type HeroName = {|
  hero: HeroName_hero
|};

// ====================================================
// GraphQL fragment: humanFragment
// ====================================================

export type humanFragment_friends = {|
  __typename: \\"Human\\",
  // What this human calls themselves
  name: string,
|} | {|
  __typename: \\"Droid\\",
  // The ID of the droid
  id: string,
|};

export type humanFragment = {|
  __typename: \\"Human\\",
  // The home planet of the human, or null if unknown
  homePlanet: ?string,
  // This human's friends, or an empty list if they have none
  friends: humanFragment_friends,
|};

// ====================================================
// GraphQL fragment: droidFragment
// ====================================================

export type droidFragment = {|
  __typename: \\"Droid\\",
  // The movies this droid appears in
  appearsIn: (?Episode)[],
|};"
`;

exports[`Flow codeGeneration fragment with fragment spreads 1`] = `
"

/* @flow */
// This file was automatically generated and should not be edited.

// ====================================================
// GraphQL fragment: simpleFragment
// ====================================================

export type simpleFragment = {|
  __typename: \\"Human\\" | \\"Droid\\",
  // The name of the character
  name: string,
|};

// ====================================================
// GraphQL fragment: anotherFragment
// ====================================================

export type anotherFragment = {|
  __typename: \\"Human\\" | \\"Droid\\",
  // The ID of the character
  id: string,
  // The name of the character
  name: string,
|};"
`;

exports[`Flow codeGeneration inline fragment 1`] = `
"

/* @flow */
// This file was automatically generated and should not be edited.

// The episodes in the Star Wars trilogy
export type Episode = \\"NEWHOPE\\" | \\"EMPIRE\\" | \\"JEDI\\";

// ====================================================
// GraphQL query operation: HeroInlineFragment
// ====================================================

export type HeroInlineFragment_hero = {|
  __typename: \\"Human\\" | \\"Droid\\",
  // The name of the character
  name: string,
  // The ID of the character
  id: string,
|};

export type HeroInlineFragment = {|
  hero: HeroInlineFragment_hero
|};"
`;

exports[`Flow codeGeneration inline fragment on type conditions 1`] = `
"

/* @flow */
// This file was automatically generated and should not be edited.

// The episodes in the Star Wars trilogy
export type Episode = \\"NEWHOPE\\" | \\"EMPIRE\\" | \\"JEDI\\";

// ====================================================
// GraphQL query operation: HeroName
// ====================================================

export type HeroName_hero_friends = {|
  __typename: \\"Human\\" | \\"Droid\\",
  // The name of the character
  name: string,
|};

export type HeroName_hero = {|
  __typename: \\"Human\\",
  // What this human calls themselves
  name: string,
  // The ID of the human
  id: string,
  // The home planet of the human, or null if unknown
  homePlanet: ?string,
  // This human's friends, or an empty list if they have none
  friends: HeroName_hero_friends,
|} | {|
  __typename: \\"Droid\\",
  // What others call this droid
  name: string,
  // The ID of the droid
  id: string,
  // The movies this droid appears in
  appearsIn: (?Episode)[],
|};

export type HeroName = {|
  hero: HeroName_hero
|};"
`;

exports[`Flow codeGeneration query with fragment spreads 1`] = `
"

/* @flow */
// This file was automatically generated and should not be edited.

// The episodes in the Star Wars trilogy
export type Episode = \\"NEWHOPE\\" | \\"EMPIRE\\" | \\"JEDI\\";

// ====================================================
// GraphQL query operation: HeroFragment
// ====================================================

export type HeroFragment_hero = {|
  __typename: \\"Human\\" | \\"Droid\\",
  // The name of the character
  name: string,
  // The ID of the character
  id: string,
|};

export type HeroFragment = {|
  hero: HeroFragment_hero
|};

// ====================================================
// GraphQL fragment: simpleFragment
// ====================================================

export type simpleFragment = {|
  __typename: \\"Human\\" | \\"Droid\\",
  // The name of the character
  name: string,
|};"
`;

exports[`Flow codeGeneration simple fragment 1`] = `
"

/* @flow */
// This file was automatically generated and should not be edited.

// ====================================================
// GraphQL fragment: SimpleFragment
// ====================================================

export type SimpleFragment = {|
  __typename: \\"Human\\" | \\"Droid\\",
  // The name of the character
  name: string,
|};"
`;

exports[`Flow codeGeneration simple hero query 1`] = `
"

/* @flow */
// This file was automatically generated and should not be edited.

// The episodes in the Star Wars trilogy
export type Episode = \\"NEWHOPE\\" | \\"EMPIRE\\" | \\"JEDI\\";

// ====================================================
// GraphQL query operation: HeroName
// ====================================================

export type HeroName_hero = {|
  __typename: \\"Human\\" | \\"Droid\\",
  // The name of the character
  name: string,
  // The ID of the character
  id: string,
|};

export type HeroName = {|
  hero: HeroName_hero
|};"
`;

exports[`Flow codeGeneration simple mutation 1`] = `
"

/* @flow */
// This file was automatically generated and should not be edited.

// The episodes in the Star Wars trilogy
export type Episode = \\"NEWHOPE\\" | \\"EMPIRE\\" | \\"JEDI\\";

// The input object sent when someone is creating a new review
type ReviewInput = {|
  stars: number,
  commentary: ?string,
  favorite_color: ?ColorInput,
|};

// The input object sent when passing in a color
type ColorInput = {|
  red: number,
  green: number,
  blue: number,
|};

// ====================================================
// GraphQL mutation operation: ReviewMovie
// ====================================================

export type ReviewMovie_createReview = {|
  __typename: \\"Review\\",
  // The number of stars this review gave, 1-5
  stars: number,
  // Comment about the movie
  commentary: ?string,
|};

export type ReviewMovie = {|
  createReview: ReviewMovie_createReview
|};"
`;
